http-client/src/main/java/io/airlift/http/client/spnego/SpnegoAuthenticationProtocolHandler.java

```diff
@@ -0,0 +1,53 @@
+package io.airlift.http.client.spnego;
+
+import org.eclipse.jetty.client.AuthenticationProtocolHandler;
+import org.eclipse.jetty.client.HttpClient;
+import org.eclipse.jetty.client.WWWAuthenticationProtocolHandler;
+import org.eclipse.jetty.client.api.Response;
+import org.eclipse.jetty.client.api.Result;
+import org.eclipse.jetty.http.HttpFields;
+import org.eclipse.jetty.http.HttpHeader;
+
+/**
+ * Jetty {@link AuthenticationProtocolHandler} requires that the
+ * {@code "WWW-Authenticate"} header contains a realm, but the
+ * {@code "Negotiate"} scheme will not have one. This hacks around that by
+ * adding a new header value for that scheme which contains a dummy realm,
+ * allowing the rest of the authentication handling to work as normal.
+ * Unfortunately, there is no easier way as all of the related code in
+ * that class is private.
+ */
+public class SpnegoAuthenticationProtocolHandler
+        extends WWWAuthenticationProtocolHandler
+{
+    private static final String NEGOTIATE = HttpHeader.NEGOTIATE.asString();
+
+    public SpnegoAuthenticationProtocolHandler(HttpClient client)
+    {
+        super(client);
+    }
+
+    @Override
+    public String getName()
+    {
+        return "spnego";
+    }
+
+    @Override
+    public Response.Listener getResponseListener()
+    {
+        return new ForwardingResponseListener(super.getResponseListener())
+        {
+            @Override
+            public void onComplete(Result result)
+            {
+                HttpHeader header = getAuthenticateHeader();
+                HttpFields headers = result.getResponse().getHeaders();
+                if (headers.getValuesList(header).stream().anyMatch(NEGOTIATE::equalsIgnoreCase)) {
+                    headers.put(header, NEGOTIATE + " realm=\"dummy\"");
```

Comment:
What if the header contains a spnego token?

Reply:
It won’t be a 401 so this code is never reached. The token is only sent on a successful authentication response.

Reply:
Add a comment explaining this.

Reply:
The Javadoc for the class explains this

Reply:
Or do you mean why it only matches without a token?

Reply:
> Or do you mean why it only matches without a token?

Yes, so that it's clear to whoever looks at this 3 months down the road since it's not "obvious" from the code.